### ISSUE TITLE
not support feign interface override？

### ISSUE THREAD

**Issue by liuliliGithub** [OpenFeign/feign#1177]
### scenes

the functional class diagram we implemented is as follows:

parent interface and child interface demo:

```java
public interface IAjRepository{
    <T extends Aj> T getAj(String bh);
    void update(String bh, String status);
}

@FeignClient("msaj")
public interface MsajRepository extends IAjRepository {  
    @GetMapping(value="/api/v1/msaj/{bh}")
    @Override
    Msaj getAj(@PathVariable("bh") String bh);

    @PatchMapping(value="/api/v1/msaj/{bh}")
    @Override
    void update(@PathVariable("bh") String bh, @RequestParam("status") String status);
}

public class AjService{
    // init data code ignore
    private Map<String, IAjRepository> ajRepositories;
    public <T extends AJ> T getAj(String bh, String ajlb) {
        return ajRepositories.get(ajlb).getAj(bh);
    }
    public void update(String bh, String ajlb, String status) {
        return ajRepositories.get(ajlb).update(bh, status);
    }
}
```

- dependency version : `10.2.3` by `spring-cloud-starter-openfeign:2.1.2.RELEASE`

### expect

`getAj()` and `update()` be invoked normally.

### actual

`update()` be invoked normally, but `geteAj()` be invoked with exception.<br>

so, not all override are not supported.

### process

with exception we found not support override because of `class.getMethods()` has 3 methods:

- `Msaj getAj(String bh)` =config-key=>`getAj(String bh)`
- `Aj getAj(String bh)` =config-key=> `getAj(String bh)`
- `void update(String bh, String status)`=config-key=>`update(String bh, String status)`

then executed code, throw exception:

```java
Util.checkState(!result.containsKey(metadata.configKey()), "Overrides unsupported: %s", new Object[]{metadata.configKey()});               
```

so, i think that if  `result.containsKey` matched, then determine whether the type of the parameter list and return is a type subclass corresponding to the currently existing metadata. after modification, it work.

```java
if(result.containsKey(metaConfigKey)) {
    MethodMetadata curMetadata = result.get(metaConfigKey);
    try{
        Class curReturnType = Class.forName(curMetadata.returnType().getTypeName());
        Class returnType = Class.forName(metadata.returnType().getTypeName());
        if(curReturnType.isAssignableFrom(returnType) || returnType.isAssignableFrom(curReturnType)) {
            continue;
        }else {
            checkState(false, "Overrides unsupported: %s",
                       metadata.configKey());
        }
    }catch(ClassNotFoundException e) {
        checkState(false, "Overrides unsupported: %s",
                   metadata.configKey());
    }
}
```

**Comment by kdavisk6** [comment]
Short Answer: This type of usage is not supported.  Please read our documentation on inheritance here: https://github.com/OpenFeign/feign#base-apis

It demonstrates how to achieve what you are looking for, by moving the annotations to the base interface and then extending them with concrete generics.  

Medium Answer: We use reflection to create a Proxy of the Target class.  As such, when dealing with inheritance, it becomes near impossible for our proxy to know which method implementation to use when a method is overridden, given that we our proxy is from an `Interface` and now a wrapper around a `Class`.  

Simply put, if you want to use inheritance, define the methods and their annotations in the root/base interface.  Interface extensions can only add more methods and not override.
